### PR TITLE
Single rendering template and tooliip

### PR DESCRIPTION
**Pull request by jonyrock**
## Problem
In this PR I am trying to avoid code repetition in templates `src/partials/template.html`, `src/directives/multibar_progress.html` and `src/directives/progress.html`, which has already showed that it's root of bugs.

### Multibar mode
Now the plugin always work in "multibar mode". All panels in "single mode"
which return more that one metric should be fixed like in example:

```sql
SELECT role, count_ok, count_warn, count_alert from multiprogress order by count_ok desc
```

should become
```sql
SELECT role, count_ok from multiprogress order by count_ok desc
```
to get previous way or showing bars.

### New labels in options
Remove of checkbox `Multibar`.

The fixes of options panel are based on https://github.com/CorpGlory/grafana-progress-list/issues/64
Now it looks like this:

image

### Directives
Merge all directives into `template.html` without using directives at all.
Angular works pretty slow and I would keep the codebase simpler.
Also removing of that template will get me closer to design of chartwerk panels.

### Tooltip
Thsi logic changed dramatically :). Now it renders faster because we update update inner contect faster and don't destroy the tooltip's html element, we just hide it. 

#### None type
image

... in case you don't want to show tooltip :)

### Options tab
* remove of "Multibar"
* validation of html in https://validator.w3.org/
* all inputs and labels aligned to two columns (see conversation below)
* filed `opacity` moved to top in the section "Coloring"

### Other
* no initialisation of directives in `Ctrl.constructor()`
* `graph_tooltip` receives data to show in a new class `TooltipItem`
* correct set of `this.items` in `Ctrl._onRender`
* there were bugs in rendering because of messy `template.html` -- not closed tag
* tooltips shows on selection of bar's labels (previously tooltip show only on bars selection, not labels)
* tooltip selection are doesn't have gaps now: you always select a progress bar if you move mouse over progresses

## Known issues
* coloring how we see it now doesn't work because it's design only for single progress bar. It fact `ProgressItem` isn't really used. With multiprogress possible to get same colors in the row  if thresholds type used.

### CODE REVIEW

File: src/partials/options.html [hunk 1]
```diff
@@ -1,327 +1,317 @@
...
+              class="gf-form-input"
+              ng-model="ctrl.panel.alias"
+              ng-change="ctrl.render()"
+              ng-model-onblur
+            />
+          </div>
+        </div>
+      </div>
     </div>
+фвыфывфывфвы
```
Comment (VargBurz) [inline src/partials/options.html:205]: Do we need it?)

File: src/partials/template.html [hunk 2]
```diff
@@ -1,20 +1,35 @@
...
+        <div
+          ng-repeat="percentValue in item.percentValues track by $index"
+          style="
+            width: {{ percentValue }}%;
+            display: inline-block;
+            background-color: {{ item.colors[$index] }};
+            opacity: {{ item.opacity }};
+            height: {{ item.titleParams.barHeight }}px;
+            margin-top: {{ item.titleParams.titleTopMargin }}px;
+          "
```
Comment (VargBurz) [inline src/partials/template.html:23]: Lets add `getItemStyle` method

File: src/mapper.ts [hunk 3]
```diff
@@ -169,7 +169,7 @@ export class ProgressItem {
   get formattedValue(): string {
     const value =
       this._panelConfig.getValue('valueLabelType') === 'percentage' ?
-      this.aggregatedProgress :
+      this.aggregatedProgress:
       this._aggregatedValue;
```
Comment (VargBurz) [inline src/mapper.ts:173]: Maybe like this?
```
const foo = maybe1 > maybe2
  ? 'bar'
  : maybeNull;
```

https://github.com/airbnb/javascript#comparison--nested-ternaries

**Review by VargBurz — COMMENTED** [review]
some code fixes and UI question.

maybe it would be better in one style?

image

File: src/graph_tooltip.ts [hunk 4]
```diff
@@ -19,16 +17,43 @@ export type Serie = {
   alias?: string
 };
 
+export class TooltipItem {
+  constructor(
+    public active: boolean, 
+    public name: string,
+    public vcs: { value: string, color: string }[]
```
Comment (rozetko) [inline src/graph_tooltip.ts:24]: it's not clear what does `vcs` mean

File: src/graph_tooltip.ts [hunk 5]
```diff
@@ -19,16 +17,43 @@ export type Serie = {
...
+          ${this.name}
+          ${this.active ? '</b>' : ''}
+        </div>
+        ${this._valuesToHtml()}
+      </div>
+    `;
+  }
+  
+  private _valuesToHtml(): string {
+    return this.vcs.map(vc => `
```
Comment (rozetko) [inline src/graph_tooltip.ts:42]: `vc` is not clear also

File: src/module.ts [hunk 6]
```diff
@@ -124,32 +118,39 @@ class Ctrl extends MetricsPanelCtrl {
...
-      () => this._seriesList, items, this.panel.tooltipMode
-    );
+    this._tooltip = new GraphTooltip(this.panel.tooltipMode);
     this._panelAlert.active = false;
 
   }
 
-  onHover(index: number, event: any, title?: any, value?: any) {
-    this._tooltip.show(event.originalEvent, index, title, value);
+  onHover(event: { index: number, event: any }, title?: any, value?: any) {
```
Comment (rozetko) [inline src/module.ts:144]:
```suggestion
  onHover(event: { index: number, event: any }) {
```
looks like we don't use them anymore

File: src/partials/options.html [hunk 7]
```diff
@@ -1,327 +1,317 @@
...
-          class="gf-form-input"
-          ng-model="ctrl.panel.statProgressType"
-          ng-options="n for n in ctrl.statProgressTypeOptions"
+      <div class="gf-form" ng-if="ctrl.panel.statProgressType == 'max value'">
+        <label class="gf-form-label width-8">Max Value</label>
+        <input
+          type="number"
+          class="gf-form-input width-14"
+          placeholder="auto" data-placement="right"
+          bs-tooltip="'Max value will be equal to max progreesss in auto mode'"
```
Comment (rozetko) [inline src/partials/options.html:61]:
```suggestion
          bs-tooltip="'Max value will be equal to max progress in auto mode'"
```

File: src/partials/options.html [hunk 8]
```diff
@@ -1,327 +1,317 @@
...
+        <h5 class="section-heading">Key Labels</h5>
+        <div class="gf-form">
+          <label class="gf-form-label width-6">Title Type</label>
+          <div class="gf-form-select-wrapper width-12">
+            <select
+              class="gf-form-input"
+              ng-model="ctrl.panel.titleViewType"
+              ng-options="n for n in ctrl.titleViewTypeOptions"
+              ng-change="ctrl.render()"
+            />
```
Comment (rozetko) [inline src/partials/options.html:182]:
```suggestion
            ></select>
```

File: src/partials/options.html [hunk 9]
```diff
@@ -1,327 +1,317 @@
...
+      <h5 class="section-heading">Coloring</h5>
+      <div class="gf-form">
+        <label class="gf-form-label width-8">Type</label>
+        <div class="gf-form-select-wrapper width-11">
+          <select
+            class="gf-form-input"
+            ng-model="ctrl.panel.coloringType"
+            ng-options="n for n in ctrl.coloringTypeOptions"
+            ng-change="ctrl.render()"
+          />
```
Comment (rozetko) [inline src/partials/options.html:221]:
```suggestion
          ></select>
```

File: src/partials/options.html [hunk 10]
```diff
@@ -1,327 +1,317 @@
...
+            <input
+              type="text" class="gf-form-input width-11"
+              ng-model="item.key"
+              ng-change="ctrl.render()" ng-model-onblur
+            />
+            <span class="gf-form-label">
+              <spectrum-picker
+                ng-model="item.color"
+                ng-change="ctrl.render()"
+              />
```
Comment (rozetko) [inline src/partials/options.html:282]:
```suggestion
              ></spectrum-picker>
```

File: src/partials/options.html [hunk 11]
```diff
@@ -1,327 +1,317 @@
...
         </div>
+
       </div>
 
     </div>
 
   </div>
-
-</div>
+</html>
```
Comment (rozetko) [inline src/partials/options.html:317]:
```suggestion
</html>

```

File: src/partials/options.html [hunk 12]
```diff
@@ -1,327 +1,317 @@
...
+            class="gf-form-input width-11"
+            ng-model="ctrl.panel.thresholds"
+            ng-blur="ctrl.render()"
+            placeholder="50,80"
           />
+        </div>
+        <div class="gf-form">
+          <label class="gf-form-label width-6">Colors</label>
+          <span class="gf-form-label">
+            <spectrum-picker ng-model="ctrl.panel.colors[0]" ng-change="ctrl.render()" />
```
Comment (rozetko) [inline src/partials/options.html:256]:
```suggestion
            <spectrum-picker ng-model="ctrl.panel.colors[0]" ng-change="ctrl.render()"></spectrum-picker>
```

File: src/partials/options.html [hunk 13]
```diff
@@ -1,327 +1,317 @@
...
+          </span>
+          <span class="gf-form-label">
+            <spectrum-picker ng-model="ctrl.panel.colors[1]" ng-change="ctrl.render()" />
```
Comment (rozetko) [inline src/partials/options.html:259]:
```suggestion
            <spectrum-picker ng-model="ctrl.panel.colors[1]" ng-change="ctrl.render()"></spectrum-picker>
```

File: src/partials/options.html [hunk 14]
```diff
@@ -1,327 +1,317 @@
...
+          </span>
+          <span class="gf-form-label">
+            <spectrum-picker ng-model="ctrl.panel.colors[2]" ng-change="ctrl.render()" />
```
Comment (rozetko) [inline src/partials/options.html:262]:
```suggestion
            <spectrum-picker ng-model="ctrl.panel.colors[2]" ng-change="ctrl.render()"></spectrum-picker>
```

File: src/partials/options.html [hunk 15]
```diff
@@ -1,327 +1,317 @@
...
-          ng-model-onblur
-        />
+        <label class="gf-form-label width-8">Sorting</label>
+        <div class="gf-form-select-wrapper width-14">
+          <select
+            class="gf-form-input"
+            ng-model="ctrl.panel.sortingOrder"
+            ng-options="n for n in ctrl.sortingOrderOptions"
+            ng-change="ctrl.render()"
+          />
```
Comment (rozetko) [inline src/partials/options.html:115]:
```suggestion
          ></select>
```

File: src/partials/options.html [hunk 16]
```diff
@@ -1,327 +1,317 @@
...
+      <h5 class="section-heading" ng-if="!ctrl.isMultibar">Tooltip</h5>
+      <div class="gf-form" ng-if="!ctrl.isMultibar">
+        <label class="gf-form-label width-8">Tooltip Mode</label>
+        <div class="gf-form-select-wrapper width-14">
+          <select
+            class="gf-form-input"
+            ng-model="ctrl.panel.tooltipMode"
+            ng-options="n for n in ctrl.tooltipModeOptions"
+            ng-change="ctrl.render()"
+          />
```
Comment (rozetko) [inline src/partials/options.html:101]:
```suggestion
          ></select>
```

File: src/partials/options.html [hunk 17]
```diff
@@ -1,327 +1,317 @@
...
         <select
-          class="gf-form-input"
-          ng-model="ctrl.panel.sortingOrder"
-          ng-options="n for n in ctrl.sortingOrderOptions"
-          ng-change="ctrl.render()"
+          class="gf-form-input width-14"
+          ng-model="ctrl.panel.mappingType"
+          ng-options="type for type in ctrl.mappingTypeOptions"
+          ng-change="ctrl.refresh()"
         />
```
Comment (rozetko) [inline src/partials/options.html:88]: can't add a suggestion here, but let's also close the tag properly :)

File: src/partials/options.html [hunk 18]
```diff
@@ -1,327 +1,317 @@
...
-        />
+      <div class="gf-form">
+        <label class="gf-form-label width-8">Progress Type</label>
+        <div class="gf-form-select-wrapper width-14">
+          <select
+            class="gf-form-input"
+            ng-model="ctrl.panel.statProgressType"
+            ng-options="n for n in ctrl.statProgressTypeOptions"
+            ng-change="ctrl.render()"
+          />
```
Comment (rozetko) [inline src/partials/options.html:51]:
```suggestion
          ></select>
```

File: src/partials/options.html [hunk 19]
```diff
@@ -1,327 +1,317 @@
...
+      <h5 class="section-heading">Mapping</h5>
+      <div class="gf-form">
+        <label class="gf-form-label width-8">Stat</label>
+        <div class="gf-form-select-wrapper width-14">
+          <select
+            class="gf-form-input"
+            ng-model="ctrl.panel.statNameOptionValue"
+            ng-options="n for n in ctrl.statNameOptions"
+            ng-change="ctrl.render()"
+          />
```
Comment (rozetko) [inline src/partials/options.html:39]:
```suggestion
          ></select>
```

File: src/partials/options.html [hunk 20]
```diff
@@ -1,327 +1,317 @@
...
-        />
+      <div class="gf-form">
+        <label class="gf-form-label width-8">Skip column</label>
+        <div class="gf-form-select-wrapper width-14">
+          <select
+            class="gf-form-input"
+            ng-model="ctrl.panel.skipColumn"
+            ng-options="n for n in ctrl.skipColumns"
+            ng-change="ctrl.render()"
+          />
```
Comment (rozetko) [inline src/partials/options.html:25]:
```suggestion
          ></select>
```

File: src/partials/options.html [hunk 21]
```diff
@@ -1,327 +1,317 @@
...
+    <div class="section gf-form-group">
+      <div class="gf-form">
+        <label class="gf-form-label width-8">Key column</label>
+        <div class="gf-form-select-wrapper width-14">
+          <select
+            class="gf-form-input"
+            ng-model="ctrl.panel.keyColumn"
+            ng-options="n for n in ctrl.columns"
+            ng-change="ctrl.render()"
+          />
```
Comment (rozetko) [inline src/partials/options.html:13]:
```suggestion
          ></select>
```

File: src/partials/options.html [hunk 22]
```diff
@@ -1,327 +1,317 @@
...
-      <label class="gf-form-label width-8">Key column</label>
-      <div class="gf-form-select-wrapper width-14">
-        <select
-          class="gf-form-input"
-          ng-model="ctrl.panel.keyColumn"
-          ng-options="n for n in ctrl.columns"
-          ng-change="ctrl.render()"
-        />
+<!DOCTYPE html>
+<html>
```
Comment (rozetko) [inline src/partials/options.html:2]: why do you wrap it in `<html>`?

**Review by rozetko — CHANGES_REQUESTED** [review]
it needs just minor changes

File: src/partials/options.html (hunk 1 above)
Comment (jonyrock) [reply src/partials/options.html:205]: It was not the task, but I agree with you and will fix it here

Comment (jonyrock) [reply src/partials/options.html]: I've made it like this:

image

**Comment by jonyrock** [comment]
Look! I returned alias label: 

image

... nobody was worried that it's not there ....

**Comment by jonyrock** [comment]
@rozetko @VargBurz I also made this:
image

File: src/partials/template.html (hunk 2 above)
Comment (jonyrock) [reply src/partials/template.html:26]: @VargBurz I like the idea! thank you

Comment (jonyrock) [reply src/partials/template.html:26]: but actually not: the thing is that you don't want to keep your logic of rendering in code as much as possible. 
It's MVP pattern here, right? So we better keep all view-related logic into view. What do you think?

File: src/partials/options.html (hunk 17 above)
Comment (jonyrock) [reply src/partials/options.html]: done, thank you

File: src/partials/options.html (hunk 22 above)
Comment (jonyrock) [reply src/partials/options.html]: it was I mistake, I tried to use https://validator.w3.org/ and that's I wrapped the file. It was required only for testing.

File: src/partials/template.html (hunk 2 above)
Comment (VargBurz) [reply src/partials/template.html:23]: yeah, it makes sense.

**Review by rozetko — APPROVED** [review]
The only thing: it would be good to know if it works in both Grafana 6 and 7

**Review by VargBurz — APPROVED** [review]

**Pull request merged**